syntax = "proto3";

import "common.proto";

service WorkerToScheduler {
    // Registers the worker with the scheduler
    rpc RegisterWorker (RegisterWorkerRequest) returns (RegisterWorkerResponse);

    // Heartbeat indicating worker status
    rpc SendHeartbeat (Heartbeat) returns (Empty);

    // Indicates to the scheduler that a job has completed
    rpc Done (DoneRequest) returns (Empty);
}

message RegisterWorkerRequest {
    string worker_type = 1;
    uint32 num_gpus = 2;
    string ip_addr = 3;
    uint32 port = 4;
}

message RegisterWorkerResponse {
    bool success = 1;
    repeated uint64 worker_ids = 2;
    uint64 round_duration = 3;
    string error_message = 4;
}

message Heartbeat {
    uint64 worker_id = 1;
    repeated JobState job_state = 2;
}

message DoneRequest {
    uint64 worker_id = 1;
    repeated uint64 job_id = 2;
    repeated uint64 num_steps = 3;
    repeated double execution_time = 4;
    repeated string iterator_log = 5;
}
